### PR TITLE
Change client sdk dependencies to versions that exist on nuget.org

### LINKED ISSUE #12164: [Bug]: `NuGet.VisualStudio` depends on package not existing on NuGet.org
### NuGet Product Used

Other/NA

### Product Version

NuGet.VisualStudio package 17.3.0 and 17.3.1

### Worked before?

_No response_

### Impact

It bothers me. A fix would be nice

### Repro Steps & Context

Reference `NuGet.VisualStudio` from NuGet.org feed.
Observe:

> Error NU1603: NuGet.VisualStudio 17.3.0 depends on Microsoft.VisualStudio.TemplateWizardInterface (>= 17.1.0-preview-2-31918-073) but Microsoft.VisualStudio.TemplateWizardInterface 17.1.0-preview-2-31918-073 was not found. An approximate best match of Microsoft.VisualStudio.TemplateWizardInterface 17.1.32210.191 was resolved.

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: https://github.com/NuGet/Home/issues/12164

Regression? probably, but it's not very impactful (just causes a NU warning on restore for anyone refernencing the package), so I'm not motivated to dig into the history to figure out when it regressed.

## Description

* Update the versions of `Microsoft.ServiceHub.Framework`,`Microsoft.VisualStudio.ComponentModelHost`, and any package using our `VSFrameworkVersion` property (such as `Microsoft.VisualStudio.SDK` and `Microsoft.VisualStudio.TemplateWizardInterface`) to versions that are available on nuget.org
* Update package source mapping config to get `Microsoft.VisualStudio.*` from nuget.org only.
  * This was necessary, because the VS SDK feed was returning HTTP 401 responses, rather than 404, causing restore to fail, even though the packages are available from nuget.org

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: Directory.Packages.props [hunk 1]
```diff
@@ -113,8 +112,8 @@
               '$(MSBuildProjectFile)' == 'NuGet.VisualStudio.Contracts.csproj' OR
               '$(MSBuildProjectFile)' == 'NuGet.VisualStudio.Interop.csproj' OR
               '$(MSBuildProjectFile)' == 'NuGet.SolutionRestoreManager.Interop.csproj'">
-    <PackageVersion Include="Microsoft.ServiceHub.Framework" Version="2.7.327-preview" />
-    <PackageVersion Include="Microsoft.VisualStudio.ComponentModelHost" Version="17.0.67-g9e37b637e6" />
+    <PackageVersion Include="Microsoft.ServiceHub.Framework" Version="4.1.112" />
```
Comment (nkolev92) [inline Directory.Packages.props:115]: I was actually working on moving this to 4.2 because there's a breaking change in 17.6, but great point about having us depend on packages that are only on nuget.org!

I'll try to account for that in my changes.

Comment (nkolev92) [reply Directory.Packages.props:115]: now I see you have a phase that validates it!

Awesome job!

File: src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore [hunk 2]
```diff
@@ -14,7 +14,6 @@ Microsoft.Internal.VisualStudio.Interop.dll
 Microsoft.Internal.VisualStudio.Shell.Framework.dll
 Microsoft.IO.Redist.dll
 Microsoft.NET.StringTools.dll
-Microsoft.ServiceHub.Client.dll
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore:17]: There should be binding redirects for this. How certain are you that we need it?

File: tools-local/ensure-nupkg-dependencies-on-source/Program.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            foreach (var (nupkgPath, packageInfo) in nupkgs)
+            {
+                foreach (var dependency in packageInfo.Dependencies)
+                {
+                    bool exists = false;
+                    if (!checkedPackages.TryGetValue(dependency, out exists))
+                    {
+                        foreach (var source in sources)
+                        {
+                            var resource = await source.GetResourceAsync<FindPackageByIdResource>();
```
Comment (nkolev92) [inline tools-local/ensure-nupkg-dependencies-on-source/Program.cs]: I'd probably generate this at line 76. You don't need to get the resource constantly.

**Review by nkolev92 — DISMISSED** [review]
LGTM. 

Suggestion for the local tool.

File: src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore (hunk 2 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore:17]: The custom code we have in the csproj that reads the `vsixignore` file was failing the build unless I removed the line. Note that the PR doesn't add it into `vsixinclude`, so we're not including it in the vsix.

I think this scenario (vsixignore lists a file that isn't in the bin directory) used to be a warning, but from memory Jeff did something to treat all msbuild warnings as errors. Maybe that's why it caused the build failure? In any case, this PR won't build unless the line is deleted.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore:17]: Great point, resolving.

**Comment by aortiz-msft** [comment]
Team review: Could we get a sample log for the new task and how long it took to run?

File: Directory.Packages.props (hunk 1 above)
Comment (zivkan) [reply Directory.Packages.props:115]: > I was actually working on moving this to 4.2 because there's a breaking change in 17.6, but great point about having us depend on packages that are only on nuget.org!

Using package dependencies on nuget.org is only important for packages we publish to nuget.org. We could theoretically use different package versions in different projects. However, the VS SDK consumption feed's 401 responses makes it difficult in practice :(   we can't (easily/feasibly) have a different nuget.config (package source mapping config) for different projects.

Anyway, point is that if we need to upgrade the VSSDK or Microsoft.ServiceHub.Framework version for our VS extension, we theoretically can. We just have to find a solution to the infrastructure issues.

**Comment by zivkan** [comment]
<details>
<summary>Sample task output</summary>
<div>

```text
2023-02-28T16:40:11.7396932Z ##[section]Starting: Check package dependencies
2023-02-28T16:40:11.7802607Z ==============================================================================
2023-02-28T16:40:11.7803284Z Task         : PowerShell
2023-02-28T16:40:11.7803474Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2023-02-28T16:40:11.7803964Z Version      : 2.212.0
2023-02-28T16:40:11.7804074Z Author       : Microsoft Corporation
2023-02-28T16:40:11.7804192Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2023-02-28T16:40:11.7804354Z ==============================================================================
2023-02-28T16:40:14.2101549Z Generating script.
2023-02-28T16:40:14.2794693Z ========================== Starting Command Output ===========================
2023-02-28T16:40:14.3225004Z ##[command]"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\49eebdca-6d76-4cbc-bc27-e9e80642b156.ps1'"
2023-02-28T16:40:14.7926123Z nupkgs:
2023-02-28T16:40:14.7991349Z D:\a\1\a\nupkgs\Microsoft.Build.NuGetSdkResolver.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.7994235Z D:\a\1\a\nupkgs\NuGet.Build.Tasks.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.7996552Z D:\a\1\a\nupkgs\NuGet.Build.Tasks.Console.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.7997749Z D:\a\1\a\nupkgs\NuGet.Build.Tasks.Pack.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8000795Z D:\a\1\a\nupkgs\NuGet.CommandLine.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8002755Z D:\a\1\a\nupkgs\NuGet.CommandLine.XPlat.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8005862Z D:\a\1\a\nupkgs\NuGet.Commands.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8008036Z D:\a\1\a\nupkgs\NuGet.Common.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8010987Z D:\a\1\a\nupkgs\NuGet.Configuration.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8013927Z D:\a\1\a\nupkgs\NuGet.Credentials.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8016458Z D:\a\1\a\nupkgs\NuGet.DependencyResolver.Core.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8017743Z D:\a\1\a\nupkgs\NuGet.Frameworks.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8020516Z D:\a\1\a\nupkgs\NuGet.Indexing.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8022732Z D:\a\1\a\nupkgs\NuGet.LibraryModel.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8023987Z D:\a\1\a\nupkgs\NuGet.Localization.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8026138Z D:\a\1\a\nupkgs\NuGet.PackageManagement.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8028093Z D:\a\1\a\nupkgs\NuGet.Packaging.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8030711Z D:\a\1\a\nupkgs\NuGet.Packaging.Core.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8033006Z D:\a\1\a\nupkgs\NuGet.ProjectModel.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8035060Z D:\a\1\a\nupkgs\NuGet.Protocol.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8037479Z D:\a\1\a\nupkgs\NuGet.Resolver.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8039623Z D:\a\1\a\nupkgs\NuGet.Versioning.6.6.0-xprivate.nupkg
2023-02-28T16:40:14.8042818Z D:\a\1\a\nupkgs\NuGet.VisualStudio.17.6.0-xprivate.nupkg
2023-02-28T16:40:14.8044086Z D:\a\1\a\nupkgs\NuGet.VisualStudio.Contracts.17.6.0-xprivate.nupkg
2023-02-28T16:40:14.8046941Z 
2023-02-28T16:40:14.8048786Z Building validation tool
2023-02-28T16:40:28.3570911Z   Determining projects to restore...
2023-02-28T16:40:56.1227758Z   Restored D:\a\1\s\tools-local\ensure-nupkg-dependencies-on-source\ensure-nupkg-dependencies-on-source.csproj (in 19.76 sec).
2023-02-28T16:41:03.5541972Z   Restored D:\a\1\s\src\NuGet.Core\NuGet.Frameworks\NuGet.Frameworks.csproj (in 27.36 sec).
2023-02-28T16:41:03.5689545Z   Restored D:\a\1\s\src\NuGet.Core\NuGet.Common\NuGet.Common.csproj (in 27.38 sec).
2023-02-28T16:41:03.5909829Z   Restored D:\a\1\s\src\NuGet.Core\NuGet.Versioning\NuGet.Versioning.csproj (in 27.36 sec).
2023-02-28T16:41:03.6393182Z   Restored D:\a\1\s\src\NuGet.Core\NuGet.Configuration\NuGet.Configuration.csproj (in 8 ms).
2023-02-28T16:41:16.0422669Z   Restored D:\a\1\s\src\NuGet.Core\NuGet.Packaging\NuGet.Packaging.csproj (in 19.88 sec).
2023-02-28T16:41:16.0428859Z   Restored D:\a\1\s\src\NuGet.Core\NuGet.Protocol\NuGet.Protocol.csproj (in 12.46 sec).
2023-02-28T16:41:28.3506114Z   NuGet.Frameworks -> D:\a\1\s\artifacts\NuGet.Frameworks\bin\Debug\netstandard2.0\NuGet.Frameworks.dll
2023-02-28T16:41:28.3534127Z   NuGet.Versioning -> D:\a\1\s\artifacts\NuGet.Versioning\bin\Debug\netstandard2.0\NuGet.Versioning.dll
2023-02-28T16:41:30.5465760Z   NuGet.Common -> D:\a\1\s\artifacts\NuGet.Common\bin\Debug\netstandard2.0\NuGet.Common.dll
2023-02-28T16:41:33.0749245Z   NuGet.Configuration -> D:\a\1\s\artifacts\NuGet.Configuration\bin\Debug\netstandard2.0\NuGet.Configuration.dll
2023-02-28T16:42:00.7127619Z   NuGet.Packaging -> D:\a\1\s\artifacts\NuGet.Packaging\bin\Debug\netcoreapp5.0\NuGet.Packaging.dll
2023-02-28T16:42:07.5177540Z   NuGet.Protocol -> D:\a\1\s\artifacts\NuGet.Protocol\bin\Debug\netcoreapp5.0\NuGet.Protocol.dll
2023-02-28T16:42:10.1947183Z   ensure-nupkg-dependencies-on-source -> D:\a\1\s\artifacts\ensure-nupkg-dependencies-on-source\bin\Debug\net7.0\ensure-nupkg-dependencies-on-source.dll
2023-02-28T16:42:10.2642325Z 
2023-02-28T16:42:10.2743608Z Build succeeded.
2023-02-28T16:42:10.2840210Z     0 Warning(s)
2023-02-28T16:42:10.2892143Z     0 Error(s)
2023-02-28T16:42:10.2942862Z 
2023-02-28T16:42:10.3290063Z Time Elapsed 00:01:47.50
2023-02-28T16:42:15.0552242Z No missing dependencies found
2023-02-28T16:42:15.1902012Z 
2023-02-28T16:42:15.1904023Z Running validation tool
2023-02-28T16:42:17.5449906Z No missing dependencies found
2023-02-28T16:42:17.8718368Z ##[section]Finishing: Check package dependencies
```

</div>
</details>

In the run of the pipeline I took the above sample output from, the new job took 3m 14s

|Task name|Duration|
|:--|--:|
|Initialize job|9s|
|Checkout NuGet/NuGet.Client|35s|
|Secure Supply Chain Analysis|6s|
|Agent Pool Validation|1s|
|DownloadBuildArtifacts|5s|
|Check package dependencies|2m 6s|
|Publish Pipeline Metadata|2s|
|Component Detection|<1s|
|Post-Job: Checkout NuGet/NuGet.Client|<1s|
|Finalize Job|<1s|

I did this as a separate job, rather than as a task in the Build_and_UnitTest job, because if we end up doing something similar for the NonRTM build's nupkgs, using a separate job won't delay the start of the mac, E2E and Apex tests. Also, if the dependencies check fails, it won't prevent the running of the tests, so the PR author will at least get feedback if their changes broke tests, rather than stopping as soon as the dependencies problem is discovered. However, since we don't have any (existing) jobs that `dependsOn` the "RTM" build, none of those precautions have any impact. On the other hand, the job runs in parallel to the Mac, E2E and Apex stages, so it doesn't increase the pipeline duration (time until pipeline is finished). But it does appear to use about 1 minute extra of CI machine time, in case CI agent costs are a concern.

File: eng/pipelines/templates/Validate_Build_Output.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+parameters:
```
Comment (nkolev92) [inline eng/pipelines/templates/Validate_Build_Output.yml]: Should we do this in a separate PR to allow cherry-picking this change to servicing branches?

**Review by nkolev92 — APPROVED** [review]

**Comment by nkolev92** [comment]
@zivkan feel free to autocomplete it. 

I can retrigger the CI if it keeps failing.
